Introduce disableToolchainDetection DSL property and fix toolchain handling in BaseNativeImageOptions

Unless I miss something, based on my tests with Liberica NIK only installed (but apply likely to GraalVM EE as well) and no GraalVM CE installation present locally (not easy because `native-build-tools` requires it to be built), even when applying `disableToolchainDetection = true` to `BuildNativeImageTask` implemented as part of #165 I get the following error:

```
> Task :nativeCompile FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':nativeCompile'.
> Error while evaluating property 'options.javaLauncher' of task ':nativeCompile'
   > Failed to query the value of property 'javaLauncher'.
      > No compatible toolchains found for request filter: {languageVersion=11, vendor=matching('GraalVM'), implementation=vendor-specific} (auto-detect true, auto-download true)
```

After a deeper analysis, I found that [toolchain relate code in `BaseNativeImageOptions`](https://github.com/graalvm/native-build-tools/blob/00a8cc3bb9755b23f7c3f8d65612885e3afa7da4/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/internal/BaseNativeImageOptions.java#L221-L228) need to be executed conditionally as well.

As a consequence, I propose to introduce `disableToolchainDetection` at DSL level (since it will be a popular option at least on Spring side for a while) to be able to just configure:
```
graalvmNative {
    disableToolchainDetection = true
}
```
And to apply it on both `BaseNativeImageOptions` and `BuildNativeImageTask` to be effective.

Be careful, to test you need to temporarily remove GraalVM CE from what is recognized by Gradle toolchain, for my tests I used Liberica NIK (available on SDKMAN) but GraalVM EE can also probably be used.

This issue is critical for Spring Native as it prevents Gradle users to use Liberica NIK which is now our default GraalVM distribution.

--- a/docs/src/docs/asciidoc/index.adoc
+++ b/docs/src/docs/asciidoc/index.adoc
@@ -20,7 +20,7 @@ If you are interested in contributing, please refer to our https://github.com/gr
 ==== Gradle plugin
 
 * Fixed resource inference not working on custom binaries
-* Fixed `disableToolchainDetection` not working if a GraalVM installation isn't present. Please use `graalvmNative.disableToolchainDetection()` instead.
+* Fixed `disableToolchainDetection` not working if a GraalVM installation isn't present. Please use `graalvmNative.toolchainDetection.set(false)` instead.
 
 === Release 0.9.8
 

--- a/docs/src/docs/asciidoc/index.adoc
+++ b/docs/src/docs/asciidoc/index.adoc
@@ -20,6 +20,7 @@ If you are interested in contributing, please refer to our https://github.com/gr
 ==== Gradle plugin
 
 * Fixed resource inference not working on custom binaries
+* Fixed `disableToolchainDetection` not working if a GraalVM installation isn't present. Please use `graalvmNative.disableToolchainDetection()` instead.
 
 === Release 0.9.8
 

--- a/docs/src/docs/snippets/gradle/groovy/build.gradle
+++ b/docs/src/docs/snippets/gradle/groovy/build.gradle
@@ -59,8 +59,8 @@ graalvmNative {
 
 if (providers.environmentVariable("DISABLE_TOOLCHAIN").isPresent()) {
 // tag::disabling-toolchain[]
-    tasks.withType(org.graalvm.buildtools.gradle.tasks.BuildNativeImageTask).configureEach {
-        disableToolchainDetection = true
+    graalvmNative {
+        disableToolchainDetection()
     }
 // end::disabling-toolchain[]
 }

--- a/docs/src/docs/snippets/gradle/groovy/build.gradle
+++ b/docs/src/docs/snippets/gradle/groovy/build.gradle
@@ -60,7 +60,7 @@ graalvmNative {
 if (providers.environmentVariable("DISABLE_TOOLCHAIN").isPresent()) {
 // tag::disabling-toolchain[]
     graalvmNative {
-        disableToolchainDetection()
+        toolchainDetection = false
     }
 // end::disabling-toolchain[]
 }

--- a/docs/src/docs/snippets/gradle/kotlin/build.gradle.kts
+++ b/docs/src/docs/snippets/gradle/kotlin/build.gradle.kts
@@ -61,7 +61,7 @@ graalvmNative {
 if (providers.environmentVariable("DISABLE_TOOLCHAIN").isPresent()) {
 // tag::disabling-toolchain[]
     graalvmNative {
-        disableToolchainDetection()
+        toolchainDetection.set(false)
     }
 // end::disabling-toolchain[]
 }

--- a/docs/src/docs/snippets/gradle/kotlin/build.gradle.kts
+++ b/docs/src/docs/snippets/gradle/kotlin/build.gradle.kts
@@ -60,8 +60,8 @@ graalvmNative {
 
 if (providers.environmentVariable("DISABLE_TOOLCHAIN").isPresent()) {
 // tag::disabling-toolchain[]
-    tasks.withType<org.graalvm.buildtools.gradle.tasks.BuildNativeImageTask>().configureEach {
-        disableToolchainDetection.set(true)
+    graalvmNative {
+        disableToolchainDetection()
     }
 // end::disabling-toolchain[]
 }

--- a/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/dsl/GraalVMExtension.java
+++ b/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/dsl/GraalVMExtension.java
@@ -85,16 +85,11 @@ public interface GraalVMExtension {
     void registerTestBinary(String name, Action<? super TestBinaryConfig> spec);
 
     /**
-     * Enables detection of toolchains which support building native images.
-     * This is the default unless {@link #disableToolchainDetection()} was called.
+     * Property driving the detection of toolchains which support building native images.
+     * The default is true.
      */
-    void enableToolchainDetection();
+    Property<Boolean> getToolchainDetection();
 
-    /**
-     * Disables detection of toolchains, which is useful when the default detection
-     * mechanism isn't capable of handling your toolchain.
-     */
-    void disableToolchainDetection();
 
     interface TestBinaryConfig {
         /**

--- a/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/dsl/GraalVMExtension.java
+++ b/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/dsl/GraalVMExtension.java
@@ -84,6 +84,18 @@ public interface GraalVMExtension {
      */
     void registerTestBinary(String name, Action<? super TestBinaryConfig> spec);
 
+    /**
+     * Enables detection of toolchains which support building native images.
+     * This is the default unless {@link #disableToolchainDetection()} was called.
+     */
+    void enableToolchainDetection();
+
+    /**
+     * Disables detection of toolchains, which is useful when the default detection
+     * mechanism isn't capable of handling your toolchain.
+     */
+    void disableToolchainDetection();
+
     interface TestBinaryConfig {
         /**
          * Sets the JVM test task which corresponds to the

--- a/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/internal/BaseNativeImageOptions.java
+++ b/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/internal/BaseNativeImageOptions.java
@@ -45,7 +45,6 @@ import org.graalvm.buildtools.gradle.dsl.NativeImageOptions;
 import org.graalvm.buildtools.gradle.dsl.NativeResourcesOptions;
 import org.graalvm.buildtools.utils.SharedConstants;
 import org.gradle.api.Action;
-import org.gradle.api.JavaVersion;
 import org.gradle.api.file.ConfigurableFileCollection;
 import org.gradle.api.model.ObjectFactory;
 import org.gradle.api.provider.ListProperty;
@@ -60,10 +59,8 @@ import org.gradle.api.tasks.Internal;
 import org.gradle.api.tasks.Nested;
 import org.gradle.api.tasks.Optional;
 import org.gradle.api.tasks.TaskContainer;
-import org.gradle.jvm.toolchain.JavaLanguageVersion;
 import org.gradle.jvm.toolchain.JavaLauncher;
 import org.gradle.jvm.toolchain.JavaToolchainService;
-import org.gradle.jvm.toolchain.JvmVendorSpec;
 
 import javax.inject.Inject;
 import java.util.Arrays;
@@ -185,6 +182,7 @@ public abstract class BaseNativeImageOptions implements NativeImageOptions {
      * to a Java launcher due to Gradle limitations.
      */
     @Nested
+    @Optional
     public abstract Property<JavaLauncher> getJavaLauncher();
 
     /**
@@ -218,14 +216,6 @@ public abstract class BaseNativeImageOptions implements NativeImageOptions {
         getSharedLibrary().convention(false);
         getImageName().convention(defaultImageName);
         getUseFatJar().convention(SharedConstants.IS_WINDOWS);
-        getJavaLauncher().convention(
-                toolchains.launcherFor(spec -> {
-                    spec.getLanguageVersion().set(JavaLanguageVersion.of(JavaVersion.current().getMajorVersion()));
-                    if (GradleUtils.isAtLeastGradle7()) {
-                        spec.getVendor().set(JvmVendorSpec.matching("GraalVM"));
-                    }
-                })
-        );
     }
 
     private static Provider<Boolean> property(ProviderFactory providers, String name) {

--- a/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/internal/DefaultGraalVmExtension.java
+++ b/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/internal/DefaultGraalVmExtension.java
@@ -45,9 +45,15 @@ import org.graalvm.buildtools.gradle.NativeImagePlugin;
 import org.graalvm.buildtools.gradle.dsl.GraalVMExtension;
 import org.graalvm.buildtools.gradle.dsl.NativeImageOptions;
 import org.gradle.api.Action;
+import org.gradle.api.JavaVersion;
 import org.gradle.api.NamedDomainObjectContainer;
 import org.gradle.api.Project;
+import org.gradle.api.provider.Property;
 import org.gradle.api.provider.Provider;
+import org.gradle.jvm.toolchain.JavaLanguageVersion;
+import org.gradle.jvm.toolchain.JavaLauncher;
+import org.gradle.jvm.toolchain.JavaToolchainService;
+import org.gradle.jvm.toolchain.JvmVendorSpec;
 
 import javax.inject.Inject;
 import java.util.HashMap;
@@ -58,15 +64,39 @@ public abstract class DefaultGraalVmExtension implements GraalVMExtension {
     private final NativeImagePlugin plugin;
     private final Project project;
     private final Map<String, Provider<Boolean>> agentProperties = new HashMap<>();
+    private final Property<JavaLauncher> defaultJavaLauncher;
+    private final JavaToolchainService toolchainService;
 
     @Inject
     public DefaultGraalVmExtension(NamedDomainObjectContainer<NativeImageOptions> nativeImages,
                                    NativeImagePlugin plugin,
-                                   Project project) {
+                                   Project project,
+                                   JavaToolchainService toolchains) {
         this.nativeImages = nativeImages;
         this.plugin = plugin;
         this.project = project;
+        this.toolchainService = toolchains;
+        this.defaultJavaLauncher = project.getObjects().property(JavaLauncher.class);
+        enableToolchainDetection();
+        nativeImages.configureEach(options -> options.getJavaLauncher().convention(defaultJavaLauncher));
         getTestSupport().convention(true);
+    }
+
+    @Override
+    public void enableToolchainDetection() {
+        defaultJavaLauncher.convention(
+                toolchainService.launcherFor(spec -> {
+                    spec.getLanguageVersion().set(JavaLanguageVersion.of(JavaVersion.current().getMajorVersion()));
+                    if (GradleUtils.isAtLeastGradle7()) {
+                        spec.getVendor().set(JvmVendorSpec.matching("GraalVM"));
+                    }
+                })
+        );
+    }
+
+    @Override
+    public void disableToolchainDetection() {
+        defaultJavaLauncher.convention((JavaLauncher) null);
     }
 
     @Override

--- a/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/internal/DefaultGraalVmExtension.java
+++ b/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/internal/DefaultGraalVmExtension.java
@@ -65,7 +65,6 @@ public abstract class DefaultGraalVmExtension implements GraalVMExtension {
     private final Project project;
     private final Map<String, Provider<Boolean>> agentProperties = new HashMap<>();
     private final Property<JavaLauncher> defaultJavaLauncher;
-    private final JavaToolchainService toolchainService;
 
     @Inject
     public DefaultGraalVmExtension(NamedDomainObjectContainer<NativeImageOptions> nativeImages,
@@ -74,7 +73,6 @@ public abstract class DefaultGraalVmExtension implements GraalVMExtension {
         this.nativeImages = nativeImages;
         this.plugin = plugin;
         this.project = project;
-        this.toolchainService = project.getExtensions().getByType(JavaToolchainService.class);
         this.defaultJavaLauncher = project.getObjects().property(JavaLauncher.class);
         getToolchainDetection().convention(true);
         nativeImages.configureEach(options -> options.getJavaLauncher().convention(defaultJavaLauncher));
@@ -86,12 +84,15 @@ public abstract class DefaultGraalVmExtension implements GraalVMExtension {
         defaultJavaLauncher.convention(
                 getToolchainDetection().flatMap(enabled -> {
                     if (enabled) {
-                        return toolchainService.launcherFor(spec -> {
-                            spec.getLanguageVersion().set(JavaLanguageVersion.of(JavaVersion.current().getMajorVersion()));
-                            if (GradleUtils.isAtLeastGradle7()) {
-                                spec.getVendor().set(JvmVendorSpec.matching("GraalVM"));
-                            }
-                        });
+                        JavaToolchainService toolchainService = project.getExtensions().findByType(JavaToolchainService.class);
+                        if (toolchainService != null) {
+                            return toolchainService.launcherFor(spec -> {
+                                spec.getLanguageVersion().set(JavaLanguageVersion.of(JavaVersion.current().getMajorVersion()));
+                                if (GradleUtils.isAtLeastGradle7()) {
+                                    spec.getVendor().set(JvmVendorSpec.matching("GraalVM"));
+                                }
+                            });
+                        }
                     }
                     return null;
                 })

--- a/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/internal/DefaultGraalVmExtension.java
+++ b/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/internal/DefaultGraalVmExtension.java
@@ -70,12 +70,11 @@ public abstract class DefaultGraalVmExtension implements GraalVMExtension {
     @Inject
     public DefaultGraalVmExtension(NamedDomainObjectContainer<NativeImageOptions> nativeImages,
                                    NativeImagePlugin plugin,
-                                   Project project,
-                                   JavaToolchainService toolchains) {
+                                   Project project) {
         this.nativeImages = nativeImages;
         this.plugin = plugin;
         this.project = project;
-        this.toolchainService = toolchains;
+        this.toolchainService = project.getExtensions().getByType(JavaToolchainService.class);
         this.defaultJavaLauncher = project.getObjects().property(JavaLauncher.class);
         enableToolchainDetection();
         nativeImages.configureEach(options -> options.getJavaLauncher().convention(defaultJavaLauncher));

--- a/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/internal/DefaultGraalVmExtension.java
+++ b/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/internal/DefaultGraalVmExtension.java
@@ -76,26 +76,26 @@ public abstract class DefaultGraalVmExtension implements GraalVMExtension {
         this.project = project;
         this.toolchainService = project.getExtensions().getByType(JavaToolchainService.class);
         this.defaultJavaLauncher = project.getObjects().property(JavaLauncher.class);
-        enableToolchainDetection();
+        getToolchainDetection().convention(true);
         nativeImages.configureEach(options -> options.getJavaLauncher().convention(defaultJavaLauncher));
         getTestSupport().convention(true);
+        configureToolchain();
     }
 
-    @Override
-    public void enableToolchainDetection() {
+    private void configureToolchain() {
         defaultJavaLauncher.convention(
-                toolchainService.launcherFor(spec -> {
-                    spec.getLanguageVersion().set(JavaLanguageVersion.of(JavaVersion.current().getMajorVersion()));
-                    if (GradleUtils.isAtLeastGradle7()) {
-                        spec.getVendor().set(JvmVendorSpec.matching("GraalVM"));
+                getToolchainDetection().flatMap(enabled -> {
+                    if (enabled) {
+                        return toolchainService.launcherFor(spec -> {
+                            spec.getLanguageVersion().set(JavaLanguageVersion.of(JavaVersion.current().getMajorVersion()));
+                            if (GradleUtils.isAtLeastGradle7()) {
+                                spec.getVendor().set(JvmVendorSpec.matching("GraalVM"));
+                            }
+                        });
                     }
+                    return null;
                 })
         );
-    }
-
-    @Override
-    public void disableToolchainDetection() {
-        defaultJavaLauncher.convention((JavaLauncher) null);
     }
 
     @Override

--- a/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/tasks/BuildNativeImageTask.java
+++ b/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/tasks/BuildNativeImageTask.java
@@ -163,7 +163,7 @@ public abstract class BuildNativeImageTask extends DefaultTask {
             logger.lifecycle("Args are: " + args);
         }
         File executablePath = null;
-        if (getDisableToolchainDetection().get()) {
+        if (getDisableToolchainDetection().get() || !options.getJavaLauncher().isPresent()) {
             if (getGraalVMHome().isPresent()) {
                 String graalvmHome = getGraalVMHome().get();
                 getLogger().lifecycle("Toolchain detection is disabled, will use GraalVM from {}.", graalvmHome);

